primitives/application-crypto/src/bls377.rs

```diff
@@ -26,3 +27,30 @@ mod app {
 #[cfg(feature = "full_crypto")]
 pub use app::Pair as AppPair;
 pub use app::{Public as AppPublic, Signature as AppSignature};
+
+impl RuntimePublic for Public {
+	type Signature = Signature;
+
+	/// Dummy implementation. Returns an empty vector.
+	fn all(_key_type: KeyTypeId) -> Vec<Self> {
+		Vec::new()
+	}
+
+	fn generate_pair(key_type: KeyTypeId, seed: Option<Vec<u8>>) -> Self {
+		sp_io::crypto::bls377_generate(key_type, seed)
```

Comment:
to my understanding, the consensus so far is to delay adding host functionality as long as possible, and instead use bls libraries directly within the runtime.
so will all other implementations here remain dummies and use runtime functions instead?
I ask because this does in fact expose bls377 key generation from the host, but this is not useful unless `sign` is implemented too. Else, one might as well keep `generate_pair` a dummy stub too.

Reply:
Unfortunately there is another code path that requires `generate_pair`.

There are a couple of places ([here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/client/rpc/src/author/mod.rs#L129-L132) and [here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/primitives/session/src/lib.rs#L132-L134)) outside the runtime which rely on the runtime (for convenience) to generate the pairs for each element in the `SessionKeys` ([this](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/bin/node-template/runtime/src/lib.rs#L87-L92) ).

For example, for the node-template:
- The runtime API implementation to generate a keypair for each element in `SessionKeys` is [this](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/bin/node-template/runtime/src/lib.rs#L418-L420) 
- `generate_session_keys` calls `generate_pair` for each `RuntimeAppPublic` in the `SessionKeys` tuple. See [here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/primitives/runtime/src/traits.rs#L2010-L2021)

Reply:
See [this comment](https://github.com/paritytech/substrate/pull/14373#issuecomment-1602043252) from @davxy. Keys are generated by the session pallet in the Runtime and so we have no easy way to customize this (and you need good randomness for that so it is not even a good idea to avoid `std` for that) . In contrast, `sign` will almost never be used in the Runtime so that should be Ok. For `verify`, because it is (currently) only being used in BEEFY pallet and we are free to have our own implementation of `verify` there therefore we could implement  the whole verify function in the runtime without involving a new host function, as it is done [here](https://github.com/w3f/substrate/blob/skalman-multiple-signature-scheme-beefy-primitves/primitives/consensus/beefy/src/lib.rs#L139C1-L139C1).

If someone else wants to use bls12-377 crypto for something else and does not know that verify is dummy then they'll be annoyed that all their signatures fail to verify till they check what is actually happening under the hood.

Reply:
on the bright side we have the `bls-experimental` feature flag for now :-)

Reply:
ack, thanks your explanations & pointing me to those uses of `generate_pair` - also don't see a workaround for this.

Reply:
> There are a couple of places ([here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/client/rpc/src/author/mod.rs#L129-L132) and [here](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/primitives/session/src/lib.rs#L132-L134)) outside the runtime which rely on the runtime (for convenience) to generate the pairs for each element in the SessionKeys ([this](https://github.com/paritytech/substrate/blob/3ef576eaeb3f42610e85daecc464961cf1295570/bin/node-template/runtime/src/lib.rs#L87-L92) ).

@davxy  Is this so that we can deprecate old session keys by no longer doing the invocation?  Or some other reason?

Reply:
I'm not aware of the original rationale of such design decision, but if I have to guess is to have some *cryptographic agility*. In this way we can add/remove pallets to the runtime which require a session key without worrying too much since the runtime is the only in charge of directly altering the keys set structure